### DISCUSSION TITLE
`isTemplateDeclaration` and `Operation` type narrowing.

### DISCUSSION THREAD

**Discussion by emi0x7d1** [microsoft/typespec#3084] [General]
We have the following [code](https://github.com/Microsoft/typespec/blob/2c47206aec8a0a2557bd2d127b05d4f95dbbc84a/packages/compiler/src/emitter-framework/type-emitter.ts#L189-L193):

```ts
    for (const operation of namespace.operations.values()) {
      if (!isTemplateDeclaration(operation)) {
        this.emitter.emitType(operation);
      }
    }
```

Inside the `if` statement, `operation` gets narrowed to `never`. This happens because `isTemplateDeclaration` is defined as:

```ts
export function isTemplateDeclaration(
  type: TemplatedType
): type is TemplatedType & { node: TemplateDeclarationNode }
```

and `Operation` is defined as:

```ts
export interface Operation extends BaseType, DecoratedType, TemplatedTypeBase {
  // ...
  node: OperationStatementNode;
  // ...
}
```

and `OperationStatementNode` as:

```ts
export interface OperationStatementNode extends TemplateDeclarationNode, /* more base types */;
```

So, `Operation` always has a `node` with type `TemplateDeclarationNode` making the `Operation` **always** a template declaration. So, typescript correctly deduces that `isTemplateDeclaration(operation)` can never be false.

I wonder what's up here. Is that `if` statement safe to remove or is something wrong else where?

Thanks in advance.

**Comment by timotheeguerin** [comment]
I think there is maybe some tweaking we should do to the templated types so that they all have a templatedeclarationode and this intersection is not needed. 

However for this specific case, I am assuming you are only looking to get template instance in that case I would use `isTemplateInsance`. For operations in particular there is a 3rd intermediate state(thought you wouldn't get it by iterating the namspace.operations). When you have a templated interface and a templated operation inside, if you instantiate the templated interface the templated operation are still not instance but not declaration either. Using `isTemplateInstance` will make sure the operation is actually an instance.
